base_report_to_printer/static/src/js/qweb_action_manager.js

```diff
@@ -0,0 +1,49 @@
+odoo.define('base_report_to_printer.print', function (require) {
+    'use strict';
+
+    var ActionManager = require('web.ActionManager');
+    var core = require('web.core');
+    var framework = require('web.framework');
+    var rpc = require('web.rpc');
+
+
+    ActionManager.include({
+        ir_actions_report: function (action, options) {
+            var action_val = _.clone(action);
+            var _t = core._t;
+            var self = this;
+            var _super = this._super;
+
+            if ('report_type' in action_val && action_val.report_type === 'qweb-pdf') {
```

Comment:
If I'm not wrong, you should be adding a new `action.report_type` called i.e. `qweb-pdf-server`, or make the action have notion of being printed in the server somehow.

Upstream code uses that key to know [if you need an HTML report](https://github.com/odoo/odoo/blob/b37cc1f9b7897759a86cdccfa1b06d58361a97da/addons/web/static/src/js/report/qwebactionmanager.js#L77) or [a PDF report](https://github.com/odoo/odoo/blob/b37cc1f9b7897759a86cdccfa1b06d58361a97da/addons/web/static/src/js/report/qwebactionmanager.js#L88), so you should just do in this method:

1. Check if `action.report_type === "qweb-pdf-server"`.
1. If it's true, do my things.
1. If it's false, `return this._super.apply(this, arguments)` and forget.

Reply:
thanks @Yajo it will take me a while to get my head around.  I don't think a new report type is correct, because every qweb-pdf report is subject to a specific or generic printing behaviour, even if that behaviour is just to send to client.

The current js works, and prints as expected, its just that it also downloads to the client, so super is being called regardless of this function.  Is there some way we can, say around line 33, once we know it is to be printed, change `action_val.report_type` so it falls through?

Reply:
Total shot in the dark, but I wonder if [Action.ir_actions_report](https://github.com/odoo/odoo/blob/6d7d6d9460cae90b257daecf848de1873c413448/addons/web/static/src/js/chrome/action_manager.js#L905). This is the first I've looked at this `js/chrome` folder though - I have no idea wtf it's for

Reply:
@lasley @Yajo OK I made massive progress.  I found a bug in the js code where action_val was checked instead of action.  As a result, none of this javascript was ever called anyway, yet it still printed the report.  

I just commented out the entire file and still got my printed report (plus the client download).

I then just manually edited the report/download http in web/controllers/main to return this (warning ugly ahead)
```python
    @http.route(['/report/download'], type='http', auth="user")
    def report_download(self, data, token):
        """This function is used by 'qwebactionmanager.js' in order to trigger the download of
        a pdf/controller report.

        :param data: a javascript array JSON.stringified containg report internal url ([0]) and
        type [1]
        :returns: Response with a filetoken cookie and an attachment header
        """
        requestcontent = json.loads(data)
        url, type = requestcontent[0], requestcontent[1]
        try:
            if type == 'qweb-pdf':
                reportname = url.split('/report/pdf/')[1].split('?')[0]
                # <-----------------snip-------------->
                error = {
                'code': 200,
                'message': "Report Sent to Printer",
                    'data': {}
                }
                return request.make_response(html_escape(json.dumps(error)))
```
and I got this

![image](https://user-images.githubusercontent.com/5093264/33964260-b664de2a-e0bc-11e7-8a9f-9442e0b7c79a.png)

so what I think the best strategy is
1. Remove the javascript
2. Override report_download with this algorithm
 - Call super
 - If it is to be sent to printer -> Return notification report has printed else return super response.

So (and sorry for using you guys as support here) how do I return a notification rather than an error in a python web controller?